### ISSUE TITLE
no-direct-mutation-state do not fail in constructor

### ISSUE THREAD

**Issue by TrustNik** [yannickcr/eslint-plugin-react#832]
Hello,
I guess following code should not be considered as a problem (now it is):
```js
export default class SomeComponent extends Component {
   constructor (props) {
           super(props);
           this.state = {a : 1};
           this.state.b = 2;
   }
...
```

What do you think?

**Comment by ljharb** [comment]
Yes, that should warn.

**Comment by TrustNik** [comment]
@ljharb could you please explain why? If I understand correctly state is initialized in the constructor, so it can't be mutated.

**Comment by ljharb** [comment]
@TrustNik it doesn't cause the same problems as mutating `this.state` elsewhere - but mutation is never a good thing.

I can see the argument that the original spirit of this rule doesn't apply to "within the constructor", but allowing mutation isn't worth making an exception imo.

**Comment by TrustNik** [comment]
@ljharb I see. Thanks for clarification.

**Comment by ivosabev** [comment]
You should never call `setState` from the constructor as it brings a lot of function calls that you don't need for a component that hasn't rendered yet. It may even potentially cause double rendering. 

The correct linting rule would be to remove the warning for `this.state` mutation and add a warning for using `setState` only when used inside the constructor.

**Comment by ljharb** [comment]
It is correct to warn about any mutation.

**Comment by ivosabev** [comment]
Except in the constructor. This behavior is against the official docs and some people think they should use the `setState` inside the constructor, when they shouldn't.

https://facebook.github.io/react/docs/state-and-lifecycle.html#do-not-modify-state-directly

**Comment by ljharb** [comment]
@ivosabev Assigning to `this.state` once in the constructor is perfectly fine. Mutating it immediately afterwards is not.

**Comment by burabure** [comment]
I have this rule Implemented here https://github.com/burabure/eslint-plugin-burabure/blob/master/lib/rules/no-direct-mutation-state.js @ljharb would you like this merged?

**Comment by ljharb** [comment]
@burabure if you have a pr with tests that can fix any broken behavior, that'd be great.

**Comment by timotgl** [comment]
@ljharb I think there is a use case for assigning `this.state` in the constructor and then mutating it immediately afterwards: when it depends on a variable number of properties, stored outside of the component. Of course you could compose the initial state object somehow (and then assign the whole thing at once), but this can get really clunky imho. (Turning an array into an object with reduce etc.)

I think disabling `no-direct-mutation-state` in this instance, for one line, can be legitimate and more readable.

**Comment by ryndshn** [comment]
was this ever fixed? i'm still receiving a warning when modifying `this.state` in the constructor.

**Comment by ljharb** [comment]
@ryndshn assigning to it once is fine; modifying it afterwards is intentionally disallowed.
